doh-server/cache.go

```diff
@@ -0,0 +1,155 @@
+package main
+
+import (
+	"encoding/json"
+	"hash/fnv"
+	"log"
+	"time"
+
+	"github.com/miekg/dns"
+)
+
+const debugCache = false
+
+func getRequestKey(msg *dns.Msg) (uint64, string) {
+	// blank out ID before creating hash
+	id := msg.Id
+	msg.Id = 0
+
+	// take away extra options, e.g. TSIG and EDNS
+	extras := msg.Extra
+	msg.Extra = nil
+
+	// use JSON to get data representation
+	var b []byte
+	var err error
+	if !debugCache {
+		b, err = json.Marshal(msg)
+	} else {
+		b, err = json.MarshalIndent(msg, "", "\t")
+	}
+	if err != nil {
+		panic(err)
+	}
+	msg.Id = id
+	msg.Extra = extras
+
+	// calculate hash
+	h := fnv.New64a()
+	h.Write(b)
+
+	if debugCache {
+		return h.Sum64(), string(b)
+	} else {
+		return h.Sum64(), ""
+	}
+}
+
+// getCached will set req.response and return true in case of cache hit;
+// return false otherwise.
+func (s *Server) getCached(req *DNSRequest) bool {
+	if !s.conf.Caching {
+		return false
+	}
+
+	req.key, req.keyJSON = getRequestKey(req.request)
+	if !debugCache {
+		req.keyJSON = ""
+	}
+
+	s.cacheLock.RLock()
+	e, ok := s.cache[req.key]
+	s.cacheLock.RUnlock()
+
+	if ok {
+		log.Printf("cache hit for request 0x%x\n", req.key)
```

Comment:
This could either be behind `Verbose` or being a new setting used to debug cache alone.

Reply:
If eventually we will have multiple "topics" to debug, I will design a better logging system so the user can choose which topic to enable.
But for now, it's OK to stick with a single `Verbose` since we don't have too many different topics.